### ISSUE TITLE
Gesture typing on bigger screen

### ISSUE THREAD

**Issue by another-sapiens** [florisboard/florisboard#579]
[You might be a bit busy, so you might wanna skip to the TLDR below]

___

___

Hi there! 
First of all, I wanted to say I absolutely love this keyboard. Packed with features, small app size, deep customization. In the FOSS world it's the most promising so far, it has pretty much all I could ask for. 

I've been following the project closely for quite some time, and the feature I'm most excited about is glide typing. Now that it's here, I'm just astonished at the amazing progress it's made in such a short amount of time. Truly amazing work!

Anyways, speaking about gesture typing, this new feature works pretty well on my Samsung A10s (even if a bit rough around the edges in terms of detecting the right words, I could perfectly type all of this). But on my Mediapad T3 7 "phablet" it won't detect pretty much anything. 
I know nothing about programming, so the only thing I can think of is that it's somehow tied to the screen size. Is that a possibility? 

Outside of that, I love this new feature.
Great job @X-yl! And thank you all for contributing to open source software with your skills.

___

___

**TLDR:**

- Phone (smaller screen): Gesture typing detects most words.
Device tested: Samsung A10s (Android 10)

- Phablet (bigger screen): Gesture typing won't detect most words.
Device tested: Huawei Mediapad T3 7 (Android 6)

If it does have to do with screen size, testing on any tablet should do, but it could be anything else. 

___

___

~~Btw, sorry if the formatting is wrong or something, this is my first post here.~~

**Comment by X-yl** [comment]
Thanks for letting me know. I have an inkling of what's going wrong. I'll investigate.

Regarding the word detection, could you perhaps try out this keyboard here: https://github.com/EtienneDesticourt/AnySoftKeyboard/tree/gesture-typing-ai (you will have to build from source) and let me know if the word detection quality is the same (I copied the algorithm from there), or whether the quality of detection in florisboard is worse, because if so I may have introduced some bugs (oops). If it's the same, then I'm afraid we'll need someone more competent at these fancy AI gizmos to improve the algorithm. One possible improvement I can think of is that we could use the next word suggestions (when they're ready) to refine the detection.

**Comment by another-sapiens** [comment]
Hi there!

I haven't built it from source, but I've used ASK in the past and, if I recall correctly, the accuracy was roughly the same as in here.
 
I think it's working as intended, though (at least on the normal-sized phone). 
What I meant before was that maybe there's a few words that confuse the algorithm (like "tied" or "like"), which you have to type manually, but overall it's good, and actually does way better at longer or more complicated words which can't get confused with similar ones.
 
I'm sure this kind of algorithm must be a real pain to make and improve. 
That said, did AnySoft's improve substantially the last two months or so? I used that one before testing new ones for a few months, and then I checked it out again and it works way better than it used to. Placebo?

Anyways, great job bringing glides to this awesome keyboard. I've permanently switched to this one some time ago and it's the only thing I missed. Thank you :)
Whether you think there's a way it could be improved through word suggestion or if we should wait for the wizards of ASK to roll out some new tweaks in the future, I'm just happy the feature's here.

**Comment by X-yl** [comment]
>I haven't built it from source, but I've used ASK in the past and, if I recall correctly, the accuracy was roughly the same as in here.

No, this is a fork of ASK which improved the gesture typing. It hasn't been merged into master so that's why you would have to build from source to try it.

> What I meant before was that maybe there's a few words that confuse the algorithm (like "tied" or "like"), which you have to type manually

Tied is one of those words that is impossible. It has the same gesture as 'tired' so there's no way for the algorithm to differentiate them (in the QWERTY layout at least. Alternate layouts that minimize collisions do exist). 'Like' should be totally possible though, and I can type it just fine. What does the keyboard detect when you try to swipe 'like'? 

Anyway, back to the big screen thing, could you give me some examples of words that can't be detected? And by "won't detect" do you mean it doesn't offer any suggestions at all? Or does it just give you the wrong word? If so, what word _does_ it detect? I'm testing on an emulated 13.5" tablet, and it seems to be alright, so possibly it's something else.

**Comment by another-sapiens** [comment]
> No, this is a fork of ASK which improved the gesture typing. It hasn't been merged into master so that's why you would have to build from source to try it.

___

Alright, I'm downloading Android Studio. Might take me a while to set it up ~~and figure out, as I'm a total noob at this~~.

___

> 'Tied' is one of those words that is impossible. It has the same gesture as 'tired' so there's no way for the algorithm to differentiate them (in the QWERTY layout at least. Alternate layouts that minimize collisions do exist). 'Like' should be totally possible though, and I can type it just fine. What does the keyboard detect when you try to swipe 'like'?

___

"Like" for some reason turns into "luke" almost every time. But there's something strange here, as one of the few words my phablet gets right is precisely that one, whereas on the phone it gets translated into "luke". 

___


> Anyway, back to the big screen thing, could you give me some examples of words that can't be detected? And by "won't detect" do you mean it doesn't offer any suggestions at all? Or does it just give you the wrong word? If so, what word does it detect? I'm testing on an emulated 13.5" tablet, and it seems to be alright, so possibly it's something else.

___

It only detects a handful of words correctly, most just type out anything else. I've just noticed something, though: If the word's first letters are too close, it doesn't detect the gliding and you can only see the letter pop up. Could that have something to do?


___

___

 
As for a few examples of the wrongly recognized words, I'll include the gliding and the result:

___


- "can" - caf

- "butcher" - bigger

- "more" - nor

- "coat" - col

- "my" - mu (or mh) 

- "doctor" - doubt

- "turn" - tut

- "enemy" eben


___

Also, some words seem to be detected correctly but missing some letters (or so it would seem). For example, gliding "there" types out "thr" sometimes, or "car" gives you "ca"


___

All these words work as intended on the phone (maybe just with an extra attempt on hard words), which is strange.


___

As for a few words that work on the phablet, we have:

- humble

- coal

- tender

- knife

- weapon (although I have to glide it as "wrapon" so it gets detected as a glide)

- heat


___

Come to think about it, there are quite a few words that work correctly, but nowhere near as many as on the phone. 

___


___

I would advice not to worry about it too much, though. My device has some weird bugs here and there, sometimes I encounter some random ones that no one else seems to have. Maybe this is one of those cases.

**Comment by another-sapiens** [comment]
I don't know what just happened, but for some very strange reason changing the option "Preview refresh delay" to 500ms did it for me, now it works perfectly! Hell, it even works better than on the phone. 

___

What does this option do, exactly?
Was it all just a giant placebo, have I finally gone mad? lol

**Comment by X-yl** [comment]
That option increases the time it takes for that preview of the word you're typing to update. How on earth that could influence anything else, I do not know. My only guess is that perhaps the high CPU usage from updating often is causing some issues or something? No idea how that could happen, though. Glad it works now lol.

> Only thing that still hasn't changed is that if first letters are too close, it won't get detected as a glide, but it's not really a problem if you extend the motion a bit more.

Hmm I'll see what I can do. The issue is if you double back on yourself, like w-e-a in weapon, you don't get far enough away for the starting point to trigger the gesture. The reason there's a cap on the time is so that when you swipe to choose a popup, it doesn't trigger as a gesture.

**Comment by another-sapiens** [comment]
> That option increases the time it takes for that preview of the word you're typing to update. How on earth that could influence anything else, I do not know. My only guess is that perhaps the high CPU usage from updating often is causing some issues or something? No idea how that could happen, though. Glad it works now lol.

That's very strange indeed. It seems to be one of those cases where "it just works" lol 
I've confirmed it on my devices, though. Lowering that option to 50ms makes a huuuuge, abismal difference in the detection. As in, it cannot get anything right, at all. Go back to 500ms and it's perfect. 

Can you try it out on your device, see if it does anything?



> Hmm I'll see what I can do. The issue is if you double back on yourself, like w-e-a in weapon, you don't get far enough away for the starting point to trigger the gesture. The reason there's a cap on the time is so that when you swipe to choose a popup, it doesn't trigger as a gesture.

Ah, it makes sense. Only way I could see it implemented is deactivating the gestures when the popup appears. But then again I know nothing about programming and have no clue on how all this stuff actually works, nor of how it could actually be implemented in practice instead of theory.

**Comment by X-yl** [comment]
My devices seem to work the same as your phone, there's no difference based on the preview delay time. Could you try the latest commit at https://github.com/X-yl/florisboard/tree/gesture-typing and see if it's fixed? The apk is [here](https://github.com/X-yl/florisboard/suites/2438688290/artifacts/52178112) if you don't want to build from source.

**Comment by another-sapiens** [comment]
Mmm, it still changes everything. Wish I could show you, but there's really a world of difference on the tablet between 50 and 500.
Maybe it is the processor or something, as even Whatsapp is laggy overall on it.
